ios/SnapyrRnSdk.m

```diff
@@ -68,8 +74,138 @@ @implementation SnapyrRnSdk
+     object:nil];
+}
+
+// Will be called when this module's last listener is removed, or on dealloc.
+-(void)stopObserving {
+    // Remove upstream listeners, stop unnecessary background tasks
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)handleSnapyrInAppMessage:(SnapyrInAppMessage *)message
+{
+    // Body of events to RN must be JSON serializable so call asDict
+    [self sendEventWithName:@"snapyrInAppMessage" body:[message asDict]];
+}
+
+- (void)handleSnapyrDidRegister:(NSNotification *)notification
+{
+    NSString *token = notification.userInfo[@"token"];
+    [self sendEventWithName:@"snapyrDidRegister" body:token];
+}
+
+- (void)handleSnapyrDidReceiveNotification:(NSNotification *)notification
+{
+    NSDictionary *notif = notification.userInfo[@"notification"];
+    [self sendEventWithName:@"snapyrDidReceiveNotification" body:notif];
+}
+
+- (void)handleSnapyrDidReceiveNotificationResponse:(NSNotification *)notification
+{
+    [self sendEventWithName:@"snapyrDidReceiveNotificationResponse" body:notification.userInfo];
+}
+
+
+- (void)handleTest:(NSNotification *)notification
+{
+    NSString *notif = notification.userInfo[@"notification"];
+    [self sendEventWithName:@"snapyrTest" body:notif];
+}
+
+// Class method the consumer can call when application(_:didRegisterForRemoteNotificationsWithDeviceToken:)
+// has been triggered on the AppDelegate.
+// Decodes token and passes it through NSNotificationCenter so it can be processed by the instance method
+// `handleSnapyrDidRegister`
++ (void)didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken
+{
+    NSUInteger dataLength = deviceToken.length;
+    if (dataLength == 0) {
+        return;
+    }
+    
+    const unsigned char *dataBuffer = (const unsigned char *)deviceToken.bytes;
+    NSMutableString *hexString = [NSMutableString stringWithCapacity:(dataLength * 2)];
+    for (int i = 0; i < dataLength; ++i) {
+        [hexString appendFormat:@"%02x", dataBuffer[i]];
+    }
+    
+    [[NSNotificationCenter defaultCenter]
+     postNotificationName:@"snapyrDidRegister"
+     object:self
+     userInfo:@{@"token" : [hexString copy]}];
+}
+
++ (void)didReceiveRemoteNotification:(NSDictionary *)notification
+{
+    [[NSNotificationCenter defaultCenter]
+     postNotificationName:@"snapyrDidReceiveNotification"
+     object:self
+     userInfo:@{@"notification": notification}];
+}
+
++ (void)didReceiveNotificationResponse:(UNNotificationResponse *)response
+API_AVAILABLE(ios(10.0))
+{
+    NSString *actionIdentifier = response.actionIdentifier;
+    NSDictionary *userInfo = [response.notification.request.content.userInfo copy];
+    [[NSNotificationCenter defaultCenter]
+     postNotificationName:@"snapyrDidReceiveNotificationResponse"
+     object:self
+     userInfo:@{@"actionIdentifier": actionIdentifier, @"userInfo": userInfo}];
+}
+
++ (BOOL)requiresMainQueueSetup
+{
+    return NO;
+}
```

Comment:
Wha?

Reply:
This is just a required method of the `RCTBridgeModule` protocol (i.e. interface). If our code needed to touch UI elements during initialization it would need to run on the main queue - but it doesn't, at least for now.